# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/Interpreter.cpp
# Commit: eeaeb40f8b86
# Full Hash: eeaeb40f8b869121eff65b1c2d11af423c9a9ecc
# Author: Iain Ireland <iireland@mozilla.com>
# Date: 2022-03-09 09:44:44
# Regressor Bug: 885514
# File Overlap Count: 1
# Description:
#   Bug 885514: Simplify JSOp::Retsub r=jandem
#   
#   Currently, JSOp::Retsub takes two arguments, and uses one to decide whether to throw the second or use it as a resume index. This patch splits out the exception-handling logic into explicit bytecode, leaving JSOp::Retsub as a simple indirect jump.
#   
#   To make the generated bytecode work out nicely, I've swapped the order of the arguments: now `throwing` is on top, and the exception/resumeindex argument is underneath it.
# ==============================================================================

diff -r 1cb0b352b024 -r eeaeb40f8b86 js/src/vm/Interpreter.cpp
--- a/js/src/vm/Interpreter.cpp	Tue Mar 08 21:10:22 2022 +0000
+++ b/js/src/vm/Interpreter.cpp	Tue Mar 08 21:10:22 2022 +0000
@@ -4105,26 +4105,11 @@
 #endif
 
     CASE(Retsub) {
-      /* Pop [exception or hole, retsub pc-index]. */
-      Value rval, lval;
-      POP_COPY_TO(rval);
-      POP_COPY_TO(lval);
-      MOZ_ASSERT(lval.isBoolean());
-      if (lval.toBoolean()) {
-        /*
-         * Exception was pending during finally, throw it *before* we adjust
-         * pc, because pc indexes into script->trynotes.  This turns out not
-         * to be necessary, but it seems clearer.  And it points out a FIXME:
-         * 350509, due to Igor Bukanov.
-         */
-        ReservedRooted<Value> v(&rootValue0, rval);
-        cx->setPendingException(v, ShouldCaptureStack::Maybe);
-        goto error;
-      }
-
-      MOZ_ASSERT(rval.toInt32() >= 0);
-
-      uint32_t offset = script->resumeOffsets()[rval.toInt32()];
+      Value resumeIndex;
+      POP_COPY_TO(resumeIndex);
+      MOZ_ASSERT(resumeIndex.toInt32() >= 0);
+
+      uint32_t offset = script->resumeOffsets()[resumeIndex.toInt32()];
       REGS.pc = script->offsetToPC(offset);
       ADVANCE_AND_DISPATCH(0);
     }
@@ -4586,7 +4571,7 @@
 
     case FinallyContinuation: {
       /*
-       * Push (true, exception) pair for finally to indicate that [retsub]
+       * Push (exception, true) pair for finally to indicate that [retsub]
        * should rethrow the exception.
        */
       ReservedRooted<Value> exception(&rootValue0);
@@ -4594,8 +4579,8 @@
         interpReturnOK = false;
         goto return_continuation;
       }
+      PUSH_COPY(exception);
       PUSH_BOOLEAN(true);
-      PUSH_COPY(exception);
       cx->clearPendingException();
     }
       ADVANCE_AND_DISPATCH(0);